### DISCUSSION TITLE
How does dependency injection work in observers?

### DISCUSSION THREAD

**Discussion by Petteroe** [Cratis/Chronicle#1467] [Q&A] [answered]
I am playing with Cratis Chronicle in an Asp.Net project, and I am a noob to both this and Event Sourcing in general. 
What I am trying to do is to have event sourcing in one module, and for some events I want to publish a domain event across boundaries using MassTransit. For this, I need to inject a MassTransit endpoint, but I get an error message:

`Cannot resolve 'EventSourcingFullExperiment.Modules.Orders.SubmitOrder.OrderSubmittedReactor' from root provider because it requires scoped service 'MassTransit.IPublishEndpoint'.`

Here is my code:

``` c#
public class OrderSubmittedReactor(IPublishEndpoint publishEndpoint, ILogger<OrderSubmittedReactor> logger) : IReactor
{
    public async Task Submitted(OrderSubmitted @event, EventContext context)
    {
        logger.LogInformation("Order {OrderId} has been submitted {@OrderSubmitted}", @event.OrderId, @event);
        
        var orderCreated = new OrderCreated(Id.FromExisting(@event.OrderId), Id.NewId(), @event.Product, @event.Quantity, @event.OrderedAtUtc);
        await publishEndpoint.Publish(orderCreated);
        
        logger.LogInformation("Order created for {OrderId} has been published {@OrderCreated}", @event.OrderId, orderCreated);
    }
}
```

Update: Just wanted to add that ILogger resolves fine when I remove the IPublishEndpoint

**Comment by woksin** [comment, marked as answer]
Hey, the problem is that the Observers wrongfully uses the root service provider and not a scoped one when resolving the services. Your IPublishEndpoint is, probably correctly, registered as a scoped service which by design cannot be resolved from the root container.

What you can do as a work around is to instead inject the IServiceProvider and in the Submitted method you can create and dispose your own service provider scope and use that to resolve your dependency. 

However as I said this is a fault in Chronicle and should be addressed

**Comment by Petteroe** [comment]
Thanks. Workaround works nicely for now :)

**Reply by woksin** [reply]
The PR #1476 should address this issue
